src/lib.rs

```diff
@@ -791,6 +815,35 @@ impl BlazeSymbolizer {
         resolver.find_line_info(addr)
     }
 
+    /// Find the addresses of a list of symbol names.
+    ///
+    /// Find the addresses of a list of symbol names from the sources
+    /// of symbols and debug info described by `sym_srcs`.
+    ///
+    /// # Arguments
+    ///
+    /// * `sym_srcs` - A list of symbol and debug sources.
+    /// * `names` - A list of symbol names.
+    pub fn find_addresses(&self, sym_srcs: &[SymbolSrcCfg], names: &[&str]) -> Vec<Vec<u64>> {
```

Comment:
let's work on API itself a bit more. Few thoughts here:

1) it seems useful sometimes to have also end address. Also when you think about uses with BPF and uprobes, what you need is not address, but actually file offset, so think about returning all that for each requested symbol

2) what do we do with inlinings of a function? this should probably be an option whether to return inlined instances or not. Results should probably also have a flag whether function was inlined at returned address or not.

3) It might be useful to also support an variant of API that accepts regexp or multiple regexps and parses all functions/symbols matching that address. In such case returning symbol name is also important

4) there are functions and there are variables, and potentially other unqualified types of objects. We probably should distinguish between them and allow querying at least functions and variables separately?

Reply:
1. Not every type of source has end addresses. You are right, it is useful sometimes. For file offsets, there are several different types of offsets for an address; for example, it also has virtual addresses, that sometimes are different from the file offset, in an ELF. We may need an API to translate addresses to various offsets.
2. It doesn't support inline functions yet. Parsing inline functions is more complicated, but we should support it at some point.  And, sure, it should be an option to create a symbolizer.
3. Regex is a complicated feature. How about providing a function to query names instead of addresses?
4. Most of the time, a caller should know a symbol is a function or variable before querying its address. I don't see the reason we should distinguish types when querying them.

Reply:
> 1. Not every type of source has end addresses. You are right, it is useful sometimes. For file offsets, there are several different types of offsets for an address; for example, it also has virtual addresses (before relocated), that sometimes are different from the file offset, in an ELF. We may need an API to translate addresses to various offsets.
> 

For things that don't have end address we can return zero. Returning file and virtaddr offsets always seems like the simplest and pretty cheap thing to do.

>     2. It doesn't support inline functions yet. Parsing inline functions is more complicated, but we should support it at some point.  And, sure, it should be an option to create a symbolizer.

I'd imagine gimli-rs supports that, we should try to utilize that library, lots of work was put into it, no need to reimplement all this on your own.

> 
>     3. Regex is a complicated feature. How about providing a function to query names instead of addresses?

Regex is complicated, but there is no need to implement it. Use excellent https://docs.rs/regex/latest/regex/

> 
>     4. Most of the time, a caller should know a symbol is a function or variable before querying its address. I don't see the reason we should distinguish types when querying them. Although, the user may need a function to return the type of a symbol. By the way, it doesn't parse the symbols of variables from DWARF. It will be added later.

Even if it's not supported right now, it's best to think how API can evolve to support conceivable use cases. If/when we allow querying static variable and functions, function/variable names won't be unique, so same name can correspond to both a function and a variable. So some sort of filter to specify whether user expects to get function information or variable information probably makes sense.

Reply:
About regex, I mean it is difficult to implement an efficient one.  We need to feed all symbol names to the state machine generated by a regex library.  How about providing an iterator to visit all symbol names?

I just added an interface to pass an additional array/Vec of parameters to extend the API in the future.

Reply:
rust's regex crate is supposed to be super fast, its author, BurntSushi, spent a lot of time on optimizing it. It powers ripgrep which is probably the fastest grep-like tool. Sure you can have slow and complicated regexes, but lots of regexes can be precompiled into very fast ones. I'd still consider supporting regex-based interface and utilize regex crate power.